### PR TITLE
Add maximum mean discrepancy and radial basis

### PR DESCRIPTION
**Pull request by kevinchern**
TODOs:

- [x] document
- [x] test
- [x] release note
- [x] wrap with `store_config`

### CODE REVIEW

**Review by kevinchern — COMMENTED** [review]
Thanks Vlad!! Nicely implemented and documented.
Let's add the MaximumMeanDiscrepancy a a module and this should be good to go.

File: dwave/plugins/torch/models/losses/mmd.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        Computes the kernel matrix for an input of shape (n, f1, f2, ...), whose shape is (n, n)
+        containing the pairwise kernel values.
+
+        Args:
+            x (torch.Tensor): A (n, f1, f2, ...) tensor.
+
+        Returns:
+            torch.Tensor: A (n, n) tensor.
+        """
+        pass
```
Comment (thisac) [inline dwave/plugins/torch/models/losses/mmd.py]: No need for pass if there is a docstring.
```suggestion
```

File: dwave/plugins/torch/models/losses/mmd.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        ``y`` are (n_x, f1, f2, ...) and (n_y, f1, f2, ...) shaped tensors, and the output is a
+        (n_x + n_y, n_x + n_y) shaped tensor containing the pairwise kernel values.
+
+        Args:
+            x (torch.Tensor): A (n_x, f1, f2, ...) tensor.
+            y (torch.Tensor): A (n_y, f1, f2, ...) tensor.
+        Returns:
+            torch.Tensor: A (n_x + n_y, n_x + n_y) tensor.
+        """
+        assert x.shape[1:] == y.shape[1:], (
```
Comment (thisac) [inline dwave/plugins/torch/models/losses/mmd.py]: Should be an if-check with an exception raised instead of an assert.

File: tests/test_dvae_winci2020.py [hunk 3]
```diff
@@ -84,6 +86,28 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
         self.dvaes = {i: DVAE(self.encoders[i], self.decoders[i]) for i in latent_dims_list}
 
+        # Now we also create a DVAE with a trainable Encoder
+        def deterministic_latent_to_discrete(logits: torch.Tensor, n_samples: int) -> torch.Tensor:
+            # straight-through estimator that maps positive logits to 1 and negative logits to -1
+            hard = torch.sign(logits)
+            soft = logits
+            result = hard - soft.detach() + soft
+            # Now we need to repeat the result n_samples times along a new dimension
+            return repeat(result, "b ... -> b n ...", n=n_samples)
```
Comment (thisac) [inline tests/test_dvae_winci2020.py]: Do we absolutely need `repeat` here? Seems a bit cumbersome to add einops as a test dependency just for this test. 🤔

**Review by thisac — COMMENTED** [review]
Thanks @VolodyaCO and @kevinchern! Test fail due to Python 3.9 tests being run (removing 3.9 support in #49, which should fix it).

Unit tests should be expanded. There should at least be test classes for kernels (`RBF`) and more unit tests for the `mmd` function. Otherwise, looks good. Just a few, mostly minor, comments.

File: dwave/plugins/torch/models/losses/mmd.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            "Input dimensions must match. You are trying to compute "
+            f"the kernel between tensors of shape {x.shape} and {y.shape}."
+        )
+        # Concatenate along batch dimension
+        xy = torch.cat([x, y], dim=0)
+        return self._kernel(xy)
+
+
+class RBFKernel(Kernel):
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply dwave/plugins/torch/models/losses/mmd.py]: Done!

File: dwave/plugins/torch/models/losses/mmd.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import torch.nn as nn
+
+from dwave.plugins.torch.nn.modules.utils import store_config
+
+__all__ = ["Kernel", "RBFKernel", "mmd_loss"]
+
+
+class Kernel(nn.Module):
+    """
+    Base class for kernels.
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply dwave/plugins/torch/models/losses/mmd.py]: Done !

File: dwave/plugins/torch/models/losses/mmd.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        """
+        Computes kernels for intra and inter set pairs in ``x`` and ``y``. In general, ``x`` and
+        ``y`` are (n_x, f1, f2, ...) and (n_y, f1, f2, ...) shaped tensors, and the output is a
+        (n_x + n_y, n_x + n_y) shaped tensor containing the pairwise kernel values.
+
+        Args:
+            x (torch.Tensor): A (n_x, f1, f2, ...) tensor.
+            y (torch.Tensor): A (n_y, f1, f2, ...) tensor.
+        Returns:
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply dwave/plugins/torch/models/losses/mmd.py]: Done

File: dwave/plugins/torch/models/losses/mmd.py (hunk 2 above)
Comment (VolodyaCO) [reply dwave/plugins/torch/models/losses/mmd.py]: Replaced it

**Comment by kevinchern** [comment]
Just did another pass. Main changes are:

1. Refactor MMD into kernels, functional, and loss modules.
2. Rename acronyms to be consistent with package standards (I'm torn between adhering to our no-acronym standard versus conventions used in pytorch, i.e., pytorch uses `MSELoss`, but we define `MaximuMeanDiscrepancyLoss`)
3. Fixed bug with RBF (need to flatten before `torch.cdist`)
4. Wrote unit tests
5. `_kernel(x)` changed to `_kernel(x, y)`.
6. Added specific errors: `class SampleSizeError(ValueError): ...` and `class DimensionMismatchError(ValueError): ...`

File: dwave/plugins/torch/nn/modules/kernels.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+class Kernel(nn.Module):
+    """Base class for kernels.
+
+    `Kernels <https://en.wikipedia.org/wiki/Kernel_method>`_ are functions that compute a similarity
+    measure between data points. Any ``Kernel`` subclass must implement the ``_kernel`` method,
+    which computes the kernel matrix for a given input multi-dimensional tensor with shape
+    (n, f1, f2, ...), where n is the number of items and f1, f2, ... are feature dimensions, so that
+    the output is a tensor of shape (n, n) containing the pairwise kernel values.
+    """
+    @abstractmethod
```
Comment (thisac) [inline dwave/plugins/torch/nn/modules/kernels.py]: Cannot use abstract methods in non-abstract classes (or rather, this won't do anything).

File: dwave/plugins/torch/nn/modules/loss.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+from dwave.plugins.torch.nn.modules.utils import store_config
+
+if TYPE_CHECKING:
+    from dwave.plugins.torch.nn.modules.kernels import Kernel
+
+__all__ = ["MaximumMeanDiscrepancyLoss"]
+
+
+class MaximumMeanDiscrepancyLoss(nn.Module):
+    """An unbiased estimator for the squared maximum mean discrepancy as a loss function.
```
Comment (thisac) [inline dwave/plugins/torch/nn/modules/loss.py]: Suggestion, since MMD is referred to in the forward method below.
```suggestion
    """An unbiased estimator for the squared Maximum Mean Discrepancy (MMD) as a loss function.
```

File: dwave/plugins/torch/nn/functional.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+class SampleSizeError(ValueError):
+    pass
+
+
+class DimensionMismatchError(ValueError):
+    pass
+
+
+def maximum_mean_discrepancy_loss(x: torch.Tensor, y: torch.Tensor, kernel: Kernel) -> torch.Tensor:
+    """Estimates the squared maximum mean discrepancy (MMD) given two samples ``x`` and ``y``.
```
Comment (thisac) [inline dwave/plugins/torch/nn/functional.py]: Shouldn't capitalized acronyms have capitalized expansions? 🤔 
```suggestion
    """Estimates the squared Maximum Mean Discrepancy (MMD) given two samples ``x`` and ``y``.
```

**Review by thisac — COMMENTED** [review]
Overall, looks good! Just a few comments/suggestions.

File: dwave/plugins/torch/nn/modules/loss.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+    This uses the ``dwave.plugins.torch.nn.functional.maximum_mean_discrepancy_loss`` function to
+    compute the loss.
+
+    Args:
+        kernel (Kernel): A kernel function object.
+    """
+
+    @store_config
+    def __init__(self, kernel: Kernel):
```
Comment (thisac) [inline dwave/plugins/torch/nn/modules/loss.py]:
```suggestion
    def __init__(self, kernel: Kernel) -> None:
```

**Review by thisac — APPROVED** [review]
